crates/bitwarden-uniffi/src/platform/fido2.rs

```diff
@@ -0,0 +1,246 @@
+        request: String,
+        client_data: ClientData,
+    ) -> Result<PublicKeyCredentialAuthenticatorAttestationResponse> {
+        let mut client = self.0 .0 .0.write().await;
+
+        let mut platform = client.platform();
+        let mut fido2 = platform.fido2();
+        let ui = UniffiTraitBridge(self.0 .1.as_ref());
+        let cs = UniffiTraitBridge(self.0 .2.as_ref());
+        let mut client = fido2.create_client(&ui, &cs)?;
+
+        let result = client.register(origin, request, client_data).await?;
+        Ok(result)
+    }
+
+    pub async fn authenticate(
+        &self,
+        origin: String,
+        request: String,
+        client_data: ClientData,
+    ) -> Result<PublicKeyCredentialAuthenticatorAssertionResponse> {
+        let mut client = self.0 .0 .0.write().await;
+
+        let mut platform = client.platform();
+        let mut fido2 = platform.fido2();
+        let ui = UniffiTraitBridge(self.0 .1.as_ref());
+        let cs = UniffiTraitBridge(self.0 .2.as_ref());
+        let mut client = fido2.create_client(&ui, &cs)?;
+
+        let result = client.authenticate(origin, request, client_data).await?;
+        Ok(result)
+    }
+}
+
+// Note that uniffi doesn't support external traits for now it seems, so we have to duplicate them
+// here.
+
+#[allow(dead_code)]
+#[derive(uniffi::Record)]
+pub struct CheckUserResult {
+    user_present: bool,
+    user_verified: bool,
+}
+
+#[uniffi::export(with_foreign)]
+#[async_trait::async_trait]
+pub trait UserInterface: Send + Sync {
+    async fn check_user(
+        &self,
+        options: CheckUserOptions,
+        credential: Option<CipherView>,
+    ) -> Result<CheckUserResult>;
+    async fn pick_credential_for_authentication(
+        &self,
+        available_credentials: Vec<Cipher>,
+    ) -> Result<CipherViewWrapper>;
+    async fn pick_credential_for_creation(
+        &self,
+        available_credentials: Vec<Cipher>,
+        new_credential: Fido2Credential,
+    ) -> Result<CipherViewWrapper>;
+}
+
+#[uniffi::export(with_foreign)]
+#[async_trait::async_trait]
+pub trait CredentialStore: Send + Sync {
+    async fn find_credentials(
+        &self,
+        ids: Option<Vec<Vec<u8>>>,
+        rip_id: String,
+    ) -> Result<Vec<Cipher>>;
+
+    async fn save_credential(&self, cred: Cipher) -> Result<()>;
+}
+
+struct UniffiTraitBridge<T>(T);
+
+#[async_trait::async_trait]
+impl bitwarden::platform::fido2::CredentialStore for UniffiTraitBridge<&dyn CredentialStore> {
+    async fn find_credentials(
+        &self,
+        ids: Option<Vec<Vec<u8>>>,
+        rip_id: String,
+    ) -> BitResult<Vec<Cipher>> {
+        self.0
+            .find_credentials(ids, rip_id)
+            .await
+            .map_err(Into::into)
+    }
+
+    async fn save_credential(&self, cred: Cipher) -> BitResult<()> {
+        self.0.save_credential(cred).await.map_err(Into::into)
+    }
+}
+
+#[derive(uniffi::Record)]
+pub struct CipherViewWrapper {
```

Comment:
Why do we need a wrapper?

Reply:
Seems to be a Uniffi limitation, which was breaking the android builds.

When a trait is the defined in the current crate but the return type is defined in an external crate, this would cause issues on the generated Kotlin code.

Using a wrapper is the best way I found to make the return type be in the same crate without creating an entire copy of `CipherView` here.

Reply:
Sounds good, lets make sure to document it though.